### PR TITLE
Improve install targets in Makefile

### PR DESCRIPTION
**Pull request by 414n**
This aims to improve a couple of shortcomings in the Makefile related to file installation system-wide (or packaging-wide) that I noticed while creating a Slackware package for this program. These are:

 - improper PREFIX variable usage, i.e. using `$(PREFIX)/usr/local` will create a `/usr/local/usr/local` or `/usr/usr/local` hierarchy depending on whether PREFIX is set to `/usr` or `/usr/local`, which is probably not desirable;
 - inconsistent use of braces instead of round brackets.

I also added a couple of lines in the README.md file to mention these adjustable parameters, but do tell me if that falls in the "too much information in the README" category.

I tested these changes by simply pointing to a DESTDIR directory under `/tmp` and checking that the files are indeed put in the expected paths:

```sh
$ ls -l /tmp/test
ls: cannot access '/tmp/test': No such file or directory
$ make DESTDIR=/tmp/test PREFIX=/usr install-autocomplete-scripts
Installing Bash auto-complete script into /tmp/test/etc/bash_completion.d/
Installing zsh auto-complete script into /tmp/test/usr/share/zsh/site-functions/
$ tree /tmp/test/
/tmp/test/
├── etc
│   └── bash_completion.d
│       └── mender-artifact
└── usr
    └── share
        └── zsh
            └── site-functions
                └── _mender-artifact

6 directories, 2 files
```

If you're okay with these changes, I'd also like to add the capability to the `install` target to install the binary directly in `$(DESTDIR)$(PREFIX)/bin` instead of the default `$GOPATH/bin`.

### CODE REVIEW

**Review by kacf — CHANGES_REQUESTED** [review]
Code wise, this looks great, it's good to have these cleaned up!

One thing needs fixing in the commit. For legal/technical reasons, the signoff string needs to match the author string exactly.

**Comment by kacf** [comment]
@lluiscampos: If we are ever in doubt in the other repositories, I believe this is how it should be done! :-)

**Comment by lluiscampos** [comment]
@414n thanks for your contribution :+1: 

> @lluiscampos: If we are ever in doubt in the other repositories, I believe this is how it should be done! :-)

@kacf Indeed. This is the way yocto defines it too.

**Comment by 414n** [comment]
> Code wise, this looks great, it's good to have these cleaned up!
> 
> One thing needs fixing in the commit. For legal/technical reasons, the signoff string needs to match the author string exactly.

Thanks, I've tried a couple of times to fix the `Signed-off-by` but did not understand what else was failing :bowing_man: 

Will now look into streamlining a bit the `install` target too.

**Review by kacf — APPROVED** [review]
Thanks! Will run this through the full test suite now.

**Comment by 414n** [comment]
I refactored a bit the `install` target too.
While doing that, I did some enhancements to the docker related targets too.
Maybe I went a bit too far, so please tell me if you prefer to revert be1d4282.

File: Makefile [hunk 1]
```diff
@@ -57,21 +73,17 @@ build-natives:
 	@env GOOS=windows GOARCH=$$arch CC=x86_64-w64-mingw32-gcc CXX=x86_64-w64-mingw32-g++ \
 		$(GO) build $(GO_LDFLAGS_WIN) $(BUILDV) -tags $(TAGS) nolzma -o $(PKGNAME)-windows.exe ;
 
-build-contained:
-	rm -f mender-artifact && \
-	image_id=$$(docker build -f Dockerfile . | awk '/Successfully built/{print $$NF;}') && \
-	docker run --rm --entrypoint "/bin/sh" -v $(shell pwd):/binary $$image_id -c "cp /go/bin/mender-artifact /binary" && \
-	docker image rm $$image_id
+build-contained: $(DOCKER_BUILD_FLAG_BASE)
```
Comment (kacf) [inline Makefile:76]: This doesn't take into account that the source might have changed. I guess it's fixable, but checking for all the source file will complicate it a bit. I guess the question is whether it's worth the hassle.

File: Makefile [hunk 2]
```diff
@@ -57,21 +73,17 @@ build-natives:
...
-	image_id=$$(docker build -f Dockerfile.binaries . | awk '/Successfully built/{print $$NF;}') && \
-	docker run --rm --entrypoint "/bin/sh" -v $(shell pwd):/binary $$image_id -c "cp /go/bin/mender-artifact* /binary" && \
-	docker image rm $$image_id
+build-natives-contained: $(DOCKER_BUILD_FLAG_NATIVES)
+	docker run --rm --entrypoint "/bin/sh" -u $$UID:$(shell id -g) -v $(shell pwd):/binary $(DOCKER_IMAGE_NAME_NATIVES) -c "cp /go/bin/$(PKGNAME)* /binary"
 
 goinstall:
 	@$(GO) install $(GO_LDFLAGS) $(BUILDV) $(BUILDTAGS)
 
+install-all: install install-autocomplete-scripts
```
Comment (kacf) [inline Makefile:85]: This should be in a separate commit.

File: Makefile [hunk 3]
```diff
@@ -67,9 +69,12 @@ build-natives-contained:
 	docker run --rm --entrypoint "/bin/sh" -v $(shell pwd):/binary $$image_id -c "cp /go/bin/mender-artifact* /binary" && \
 	docker image rm $$image_id
 
-install:
+goinstall:
 	@$(GO) install $(GO_LDFLAGS) $(BUILDV) $(BUILDTAGS)
 
+install: $(FINALEXE)
+	install -Dm0755 $(FINALEXE) $(DESTDIR)$(PREFIX)/bin/$(FINALEXE)
```
Comment (kacf) [inline Makefile:88]: We've had bad experience with this in the past, see [this commit](https://github.com/mendersoftware/mender/commit/b6e20f1f274e55fc09c9b0c09b5c8f9429ca2b05). `go install` caches the build artifacts, so I don't think this is very important for performance. Unless there's a strong reason I think it's better to keep this the same across the Mender repositories.

**Review by kacf — CHANGES_REQUESTED** [review]
FYI, I merged the commits we already agreed on.

I'm a bit more sceptical of the new commits. Take a look at the comments and we can take it from there.

File: Makefile (hunk 1 above)
Comment (414n) [reply Makefile:76]: Yes, you're right, at the moment it only looks for changes in the related Dockerfile.
I'm still looking into making it aware of Go source file changes too but, as you said, it probably won't be easy as Go projects structure is not really Makefile-friendly. If you deem this as overcomplicated/overengineered, I can leave it as that by reverting the last commit.

File: Makefile (hunk 2 above)
Comment (414n) [reply Makefile:85]: I'm sorry, it should've went into 3733b77bcc8d01ed7e5be2c7f73e1131e1f84932.
Will fix it ASAP

File: Makefile (hunk 3 above)
Comment (414n) [reply Makefile:88]: I think there's a misunderstanding here. I changed the `install` target mainly to map it to the action of installing the final binary in a standard system-wide location (`$PREFIX/bin`, optionally prefixed with a `$DESTDIR` path useful when packaging), while before it was just called to build the final binary and put it under `$GOPATH/bin`.
Now, by looking at mendersoftware/mender@b6e20f1f2, I can see the concern regarding build tags change detection that can be mismanaged by only adding a target dependency on `FINALEXE`.
I can definitely look into it.

File: Makefile (hunk 1 above)
Comment (kacf) [reply Makefile:76]: Yes, given that the benefit is pretty low (Go already caches build artifacts), I think I would vote for omitting this to keep the complexity down.

File: Makefile (hunk 3 above)
Comment (kacf) [reply Makefile:88]: Yes, the Go command would not obey `PREFIX` and `DESTDIR`.

I think this can be solved by simply replacing the dependency on `$(FINALEXE)` with `build` instead, and then making sure that `build` does not depend on it either (to avoid short circuiting the build).

File: Makefile (hunk 2 above)
Comment (414n) [reply Makefile:85]: Force pushed the change just now.
Sorry it took so long.

File: Makefile (hunk 3 above)
Comment (414n) [reply Makefile:88]: Removed the `$(FINALEXE)` dependency in favor of the phony `build` target dependency: this should enforce a fresh build on every `install` invocation.

File: Makefile (hunk 1 above)
Comment (kacf) [reply Makefile:76]: This commit should be removed, as per the discussion.

There is definitely room for improvement here, but we can't submit something that doesn't take code changes into account.

**Review by kacf — CHANGES_REQUESTED** [review]
The first and last commit are ready to go as-is. I suggest you just remove the "enhance docker targets", so we can get those two commits merged. If you still want to work on "enhance docker targets", it can be done separately.

**Comment by 414n** [comment]
Ok, got rid of the docker targets "enhancements" and kept only the other two commits, as suggested.

Let me know if anything needs further polishing.

**Review by kacf — APPROVED** [review]
Sorry for the delay. Yes, it looks good now! I'll run this through our test framework.
